### PR TITLE
gh-126775: make linecache.checkcache threadsafe and GC re-entrency safe

### PR DESCRIPTION
**Pull request by graingert**
* Issue: gh-126775

### CODE REVIEW

File: Misc/NEWS.d/next/Library/2024-11-13-10-44-25.gh-issue-126775.a3ubjh.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+make :func:`linecache.checkcache` threadsafe and GC re-entrency safe
```
Comment (johnslavik) [inline Misc/NEWS.d/next/Library/2024-11-13-10-44-25.gh-issue-126775.a3ubjh.rst]:
```suggestion
Make :func:`linecache.checkcache` thread safe and GC re-entrancy safe.
```

File: Lib/linecache.py [hunk 2]
```diff
@@ -49,14 +49,16 @@ def checkcache(filename=None):
     (This is not checked upon each call!)"""
 
     if filename is None:
-        filenames = list(cache.keys())
-    elif filename in cache:
-        filenames = [filename]
+        filenames = cache.copy().keys()
```
Comment (cfbolz) [inline Lib/linecache.py:53]: I would add a comment here why the copy is necessary. Apart from that it looks good to me.

**Review by cfbolz — CHANGES_REQUESTED** [review]

File: Lib/linecache.py (hunk 2 above)
Comment (graingert) [inline Lib/linecache.py:53]:
```suggestion
        # get keys atomically
        filenames = cache.copy().keys()
```

**Comment by graingert** [comment]
I have made the requested changes; please review again

**Review by cfbolz — APPROVED** [review]

**Pull request merged**